docs/proposals/31-inplace-node-update.md

```diff
@@ -0,0 +1,389 @@
+        minimum: 5
+        maximum: 5
+        maxSurge: 0
+        maxUnavailable: 2
+        updateStrategy: InPlaceUpdate # RollingUpdate/InPlaceUpdate/InplaceUpdateOnLabel
+        machine:
+          type: m5.large
+          image:
+            name: <some-image-name>
+            version: <some-image-version>
+            providerConfig: <some-machine-image-specific-configuration>
+          architecture: <some-cpu-architecture>
+  kubernetes:
+    version: 1.27.3
+  networking:
+    type: <some-network-extension-name> # {calico,cilium}
+```
+
+#### `Worker` API
+
+A new field `updateStrategy` is introduced under `.spec.pools[]` in the Worker API. This field will be used to set the `machineDeploymentStrategy` for the `machinev1alpha1.MachineDeployment` resource.
+
+```yaml
+apiVersion: extensions.gardener.cloud/v1alpha1
+kind: Worker
+metadata:
+  name: worker
+  namespace: shoot--foobar--aws
+spec:
+  type: aws
+  region: eu-west-1
+  secretRef:
+    name: cloudprovider
+    namespace: shoot--foobar--aws
+  infrastructureProviderStatus: <infra-status>
+  pools:
+    - name: cpu-worker-1
+      machineType: m4.large
+      machineImage:
+        name: gardenlinux
+        version: 1.2.2
+      minimum: 5
+      maximum: 5
+      maxSurge: 0
+      maxUnavailable: 2
+      updateStrategy: InPlaceUpdate # RollingUpdate/InPlaceUpdate/InplaceUpdateOnLabel
+```
+
+#### `OperatingSystemConfig` API
+
+A new field `version` is introduced in the `OperatingSystemConfig` spec. It specifies the machine image/OS version of the worker pool.
+
+```yaml
+apiVersion: extensions.gardener.cloud/v1alpha1
+kind: OperatingSystemConfig
+metadata:
+  name: pool-01-original
+  namespace: default
+spec:
+  type: <os-type>
+  version: <operating-system-version>
+  units:
+    - ...
+  files:
+    - ...
+```
+
+#### Gardener Node Agent
+
+The `gardener-node-agent` will function as the tool responsible for updating Kubernetes minor version and Machine image version dependencies on the node.
+
+Currently, when a dependency changes in the Operating System Configuration (OSC), `gardener-node-agent` detects the difference between the current OSC and the last applied OSC, and updates any changed unit on the node. When the in-place update strategy is configured, and either the Kubernetes minor version or the machine image version is changed, `gardener-node-agent` will not immediately apply the updates to the node. Instead, it will wait for the `ready-to-update` label on the node. Once the label is present, `gardener-node-agent` will proceed with the update.
+
+In the case of OS updates, `gardener-node-agent` expects an "update-script" to be present in the `.spec.files[]` of the OSC at a predefined path, and the same will be executed to perform the update.
+
+After the update, based on its success or failure, `gardener-node-agent` will add the appropriate label(`update-successful` or `update-failed`) to the node to indicate the status of the update.
```

Comment:
When will these labels be removed?

Reply:
Request that `update-successful` have a normative, information message for the label value that is logable by the MCM and be capable of being put in the `Machine` status.

Reply:
MCM removes all the labels after the update is successful and the machine is moved to the new machine set.